flink-filesystems/flink-gs-fs-hadoop/src/main/java/org/apache/flink/fs/gs/utils/ConfigUtils.java

```diff
@@ -0,0 +1,202 @@
+        // 3) use Hadoop "google.cloud.auth.service.account.json.keyfile" as location of
+        // credentials, if supplied
+        // 4) use no credentials
+
+        // store any credentials we are to use, here
+        Optional<String> credentialsPath = Optional.empty();
+
+        // only look for credentials if service account support is enabled
+        boolean enableServiceAccount =
+                hadoopConfig.getBoolean("google.cloud.auth.service.account.enable", true);
+        if (enableServiceAccount) {
+
+            // load google application credentials, and then fall back to
+            // "google.cloud.auth.service.account.json.keyfile" from Hadoop
+            credentialsPath = configContext.getenv("GOOGLE_APPLICATION_CREDENTIALS");
+            if (credentialsPath.isPresent()) {
+                LOGGER.info(
+                        "GSRecoverableWriter is using GOOGLE_APPLICATION_CREDENTIALS at {}",
+                        credentialsPath.get());
+            } else {
+                credentialsPath =
+                        Optional.ofNullable(
+                                hadoopConfig.get("google.cloud.auth.service.account.json.keyfile"));
+                credentialsPath.ifPresent(
+                        path ->
+                                LOGGER.info(
+                                        "GSRecoverableWriter is using credentials from Hadoop at {}",
+                                        path));
+            }
+        }
+
+        // construct the storage options
+        StorageOptions.Builder storageOptionsBuilder = StorageOptions.newBuilder();
+        if (credentialsPath.isPresent()) {
+            LOGGER.info(
+                    "Creating GSRecoverableWriter using credentials from {}",
+                    credentialsPath.get());
+            configContext.setStorageCredentialsFromFile(
+                    storageOptionsBuilder, credentialsPath.get());
+        } else {
+            LOGGER.info("Creating GSRecoverableWriter using no credentials");
+        }
+
+        return storageOptionsBuilder.build();
+    }
+
+    /**
+     * Helper to serialize a Hadoop config to a string, for logging.
+     *
+     * @param hadoopConfig The Hadoop config.
+     * @return A string with the Hadoop properties.
+     * @throws RuntimeException On underlying IO failure
+     */
+    public static String stringifyHadoopConfig(org.apache.hadoop.conf.Configuration hadoopConfig)
+            throws RuntimeException {
+        try (Writer writer = new StringWriter()) {
+            org.apache.hadoop.conf.Configuration.dumpConfiguration(hadoopConfig, writer);
+            return writer.toString();
+        } catch (IOException ex) {
+            throw new RuntimeException(ex);
+        }
+    }
+
+    /**
+     * Interface that provides context-specific config helper functions, factored out to support
+     * unit testing. *
+     */
+    public interface ConfigContext {
```

Comment:
I find a bit hard to understand the responsibility of this interface. It seems several things are mixed together.
1. It serves as a provider of context-related inputs: environment and files.
2. It somehow also decides how context-related inputs are applied: overwriting the given config / storage options.
3. In the tests, it's also used for recording which files the util is reading.

I think 1) alone should be good enough for providing different context-related inputs in production / tests. 2) is probably not a big deal as the logics are as simple as passing the around. However, I'm not sure with 3) as it feels like we are checking that the `ConfigUtils` is reading the correct input rather than providing the correct output.

Reply:
Another thing that doesn't feel right is having to provide different `ConfigContext` implementations for various test cases. And the `UnsupportedOperationException` indicates that we are assuming how `ConfigContext` is used by `ConfigUtils` internally, rather than treat the latter as a blackbox.

Reply:
I would suggest the following:

1. Change the `ConfigContext` protocol to a mere input provider. Something like:
```
public interface ConfigContext {
    Optional<String> getenv(String name);
    org.apache.hadoop.conf.Configuration loadHadoopConfigFromDir(String configDir);
    GoogleCredentials loadStorageCredentialsFromFile(String credentialsPath);
}
```

2. We can have a testing implementation like:
```
public class TestingConfigContext implements ConfigContext{
    Map<String, String> envs;
    Map<String, org.apache.hadoop.conf.Configuration> hadoopConfigs;
    Map<String, GoogleCredentials> credentials;

    public Optional<String> getenv(String name) {
        return Optional.ofNullable(envs.get(name));
    }
    public org.apache.hadoop.conf.Configuration loadHadoopConfigFromDir(String configDir) {
        return hadoopConfigs.get(configDir);
    }
    public GoogleCredentials loadStorageCredentialsFromFile(String credentialsPath) {
        return credentials.get(credentialsPath);
    }
}
```
In this way we can reuse the same `TestingConfigContext` implementation in different test cases, constructed with different parameters.

3. The test cases can simply verify the outcome of `ConfigUtils#getHadoopConfiguration` and `ConfigUtils#getStorageOptions` (`StorageOptions#getCredentials`).

Reply:
Good suggestions, done in https://github.com/apache/flink/pull/18489/commits/c258789c884f9a5733c59e53f213ae635520a1e3.

I had to make one more change. It turns out that, if one creates a `StorageOptions` instance via `StorageOptions.Builder#build` without having provided credentials to the builder, the builder will use credentials defined via `GOOGLE_APPLICATION_CREDENTIALS` -- and return them via `getCredentials` -- and there doesn't seem to be a way to prevent this. This isn't a problem at runtime, but it does create problems for the tests if `GOOGLE_APPLICATION_CREDENTIALS` is defined in the environment; specifically, `StorageOptions` that we would expect to have no credentials via `getCredentials` in fact do have credentials defined.

So, I changed `ConfigUtils#getStorageOptions` to be `ConfigUtils#getStorageCredentials` instead, i.e.:

```
    public static Optional<GoogleCredentials> getStorageCredentials(
            org.apache.hadoop.conf.Configuration hadoopConfig, ConfigContext configContext) {
```

... which allows me to properly validate in unit tests whether credentials were created or not, avoiding the need to read them back out of a `StorageOptions` instance.